core/src/main/java/org/geysermc/geyser/translator/protocol/bedrock/BedrockNetworkStackLatencyTranslator.java

```diff
@@ -46,47 +38,18 @@ public class BedrockNetworkStackLatencyTranslator extends PacketTranslator<Netwo
 
     @Override
     public void translate(GeyserSession session, NetworkStackLatencyPacket packet) {
-        // negative timestamps are used as hack to fix the url image loading bug
-        if (packet.getTimestamp() >= 0) {
-            if (session.getGeyser().getConfig().isForwardPlayerPing()) {
-                // use our cached value because
-                // a) bedrock can be inaccurate with the value returned
-                // b) playstation replies with a different magnitude than other platforms
-                // c) 1.20.10 and later reply with a different magnitude
-                Long keepAliveId = session.getKeepAliveCache().poll();
-                if (keepAliveId == null) {
-                    session.getGeyser().getLogger().debug("Received a latency packet that we don't have a KeepAlive for: " + packet);
-                    return;
-                }
-
-                ServerboundKeepAlivePacket keepAlivePacket = new ServerboundKeepAlivePacket(keepAliveId);
-                session.sendDownstreamPacket(keepAlivePacket);
-            }
+        // We should receive these packets in the same order they were sent
+        final Runnable latencyPing = session.getLatencyPingCache().poll();
+        if (latencyPing == null) {
+            session.getGeyser().getLogger().debug("Received a latency packet that we don't have a ping for: " + packet);
             return;
         }
 
-        if (session.getPendingOrCurrentBedrockInventoryId() != -1) {
-            InventoryUtils.openPendingInventory(session);
-        } else {
-            session.scheduleInEventLoop(() -> {
-                // Hack to fix the url image loading bug
-                UpdateAttributesPacket attributesPacket = new UpdateAttributesPacket();
-                attributesPacket.setRuntimeEntityId(session.getPlayerEntity().getGeyserId());
-
-                AttributeData attribute = session.getPlayerEntity().getAttributes().get(GeyserAttributeType.EXPERIENCE_LEVEL);
-                if (attribute != null) {
-                    attributesPacket.setAttributes(Collections.singletonList(attribute));
-                } else {
-                    attributesPacket.setAttributes(Collections.singletonList(GeyserAttributeType.EXPERIENCE_LEVEL.getAttribute(0)));
-                }
-
-                session.sendUpstreamPacket(attributesPacket);
-            }, 500, TimeUnit.MILLISECONDS);
-        }
+        latencyPing.run();
     }
 
     @Override
     public boolean shouldExecuteInEventLoop() {
```

Comment:
Nitpick - we can remove this function now since the parent method is true.

Reply:
Actually... maybe we should revert this and instead make Runnables in latencyPingCache ensure we're in the session loop? Doing this will increase the ping Java players see in the tab list.

Reply:
I don't follow. As per above, these packets are sync, not async

Reply:
`shouldExecuteInEventLoop` controls if *we* treat these packets as sync or async within our event loop. We should probably treat all packets as async, or else there will be a further, tiny delay when dealing with keep alive packets that are async. For NetworkStackLatencyPackets that should be treated as sync on both Java and Bedrock, we can just add them to our event loop from the called runnable. Is that correct?